pkg/rbac/userData.go

```diff
@@ -0,0 +1,147 @@
+// Copyright Contributors to the Open Cluster Management project
+package rbac
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/stolostron/search-v2-api/pkg/config"
+	authzv1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	v1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+// Contains data about the resources the user is allowed to access.
+type userData struct {
+	// clusters     []string              // Managed clusters where the user has view access.
+	// csResources  []resource            // Cluster-scoped resources on hub the user has list access.
+	nsResources map[string][]resource // Namespaced resources on hub the user has list access.
+
+	// Internal fields to manage the cache.
+	// clustersErr       error      // Error while updating clusters data.
+	// clustersLock      sync.Mutex // Locks when clusters data is being updated.
+	// clustersUpdatedAt time.Time  // Time clusters was last updated.
+	err     error      // Error while getting user data from cache
+	csrErr  error      // Error while updating cluster-scoped resources data.
+	csrLock sync.Mutex // Locks when cluster-scoped resources data is being updated.
+	// csrUpdatedAt time.Time  // Time cluster-scoped resources was last updated.
+	nsrErr       error      // Error while updating namespaced resources data.
+	nsrLock      sync.Mutex // Locks when namespaced resources data is being updated.
+	nsrUpdatedAt time.Time  // Time namespaced resources was last updated.
+
+	// impersonate *kubernetes.Interface // client with impersonation config
+}
+
+func (cache *Cache) GetUserData(ctx context.Context, clientToken string) (*userData, error) {
+	var user userData
+	uid := cache.tokenReviews[clientToken].tokenReview.Status.User.UID //get uid from tokenreview
+	cache.usersLock.Lock()
+	defer cache.usersLock.Unlock()
+	cachedUserData, userDataExists := cache.users[uid] //check if userData cache for user already exists
+	if userDataExists {
+		klog.V(5).Info("Using user data from cache.")
+		return cachedUserData, user.err
+
+	}
+	// create new instance
+	cache.users[uid] = &user
```

Comment:
Looks like a logic error - we are assigning nil value , should the userData from next line . @Anxhela21